File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.py
# 조합?   탐욕법?

# 1. 조합
    # 5C3 일때 sum(arr) == 0 인것 고르는 것인가

import sys
from itertools import combinations as c
N, S = list(map(int, input().strip().split()))
arr = list(map(int, input().strip().split()))

cnt = 0
for i in range(1, N+1):
    comb = c(arr, i)

    for x in comb:
        if sum(x) == S:
            cnt += 1
print(cnt)

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {
	static int N, S;
	static int cnt = 0;
	static int[] arr;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String[] NS = br.readLine().split(" ");
		String[] s_arr = br.readLine().split(" ");
		N = Integer.parseInt(NS[0]);
		S = Integer.parseInt(NS[1]);
		arr = new int[N];
		
		for (int i = 0; i < N; i++) {
			arr[i] = Integer.parseInt(s_arr[i]);
		}
		
		// 1. 초기 탐색 시작
		for (int i = 0; i < N; i++) {
			back_tracking(i+1, arr[i]);
		}
		
		// 3. 출력
		System.out.println(cnt);
	}
	
	

	// 2. 조합 (부분집합의 합을 구하기 위함)
	static void back_tracking(int depth, int sum) {
		
		// 2-1. 최대깊이 도달시
		if (depth == N) {
			if (sum == S) cnt++;
			return;
		}
		
		// 2-2. 아직 도달 못했을때
		
		// a. 선택
		back_tracking(depth + 1, sum + arr[depth]);
		
		// b. 미선택
		back_tracking(depth + 1, sum);
	}
	
	
}